File: packages/plugin-sdk-go/aisoc/action.go
package aisoc

import "context"

// ActionRequest is the payload sent to a response-action plugin.
type ActionRequest struct {
	ActionID      string         `json:"action_id"`
	Params        map[string]any `json:"params,omitempty"`
	DryRun        bool           `json:"dry_run"`
	CaseID        string         `json:"case_id,omitempty"`
	PlaybookRunID string         `json:"playbook_run_id,omitempty"`
}

// ActionResult is the result returned by a response-action plugin.
type ActionResult struct {
	ActionID string         `json:"action_id"`
	Success  bool           `json:"success"`
	DryRun   bool           `json:"dry_run"`
	Summary  string         `json:"summary,omitempty"`
	Details  map[string]any `json:"details,omitempty"`
	Error    string         `json:"error,omitempty"`
}

// Action is implemented by plugins that perform automated response actions.
type Action interface {
	Plugin
	// SupportedActions returns the list of action IDs this plugin handles.
	SupportedActions() []string
	// Execute runs the action and returns a structured result.
	Execute(ctx context.Context, req ActionRequest, pctx PluginContext) (ActionResult, error)
}

File: packages/plugin-sdk-py/src/aisoc_plugin_sdk/action.py
"""Response-action plugin base class."""

from __future__ import annotations

from abc import abstractmethod
from typing import Any

from pydantic import BaseModel, Field

from .plugin import AiSOCPlugin, PluginContext


class ActionRequest(BaseModel):
    """Payload sent to a response-action plugin."""

    action_id: str = Field(..., description="Registered action identifier")
    params: dict[str, Any] = Field(default_factory=dict, description="Action-specific parameters")
    dry_run: bool = Field(
        False,
        description="If True the plugin must validate params and return what would happen without executing",
    )
    case_id: str | None = Field(None, description="Associated case ID, if any")
    playbook_run_id: str | None = Field(None, description="Playbook run that triggered this action")


class ActionResult(BaseModel):
    """Result returned by a response-action plugin."""

    action_id: str
    success: bool
    dry_run: bool = False
    summary: str = Field(default="", description="Human-readable one-liner of what happened")
    details: dict[str, Any] = Field(default_factory=dict, description="Structured action output")
    error: str | None = None


class ActionPlugin(AiSOCPlugin):
    """Base class for response-action plugins."""

    @abstractmethod
    def supported_actions(self) -> list[str]:
        """Return list of action IDs this plugin handles."""

    @abstractmethod
    async def execute(self, request: ActionRequest, ctx: PluginContext) -> ActionResult:
        """Execute the action and return a structured result."""
